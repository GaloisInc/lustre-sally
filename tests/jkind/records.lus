-- This example exercises the flattening of records
--
-- Expected output:
-- VALID PROPERTIES: [lemma, ok1]
-- INVALID PROPERTIES: [cex1, cex2] (K = 11)

type point = struct { x : int; y : int };
type weighted_point = struct { weight : real; p : point };

type large = struct { a : int; b : int; c : int; d : int; e : int };

node main(delta1, delta2 : subrange [-1, 1] of int) returns ();
var
  wp1, wp2 : weighted_point;
  lg : large;
  ok1, cex1, cex2, lemma : bool;
let
  wp1 = weighted_point {
          weight = 1.0 -> pre(wp2.weight);
          p = point {
            x = 0 -> pre(wp1.p.x) + 1;
            y = 0 -> pre(wp1.p.y) + delta1}};
  wp2 = weighted_point {
          weight = 2.0 -> pre(wp1.weight);
          p = point {
            x = 0 -> pre(wp2.p.x) + 1;
            y = 20 -> pre(wp2.p.y) + delta2}};

  cex1 = not (wp1.p = wp2.p);
  --%PROPERTY cex1;

  -- This lemma is required for ok1
  lemma = wp1.weight <> wp2.weight;
  --%PROPERTY lemma;
  
  ok1 = wp1 <> wp2;
  --%PROPERTY ok1;

  lg = large { a = 0; b = 0; c = 0; d = 0; e = 0 } ->
         if delta1 = 0 then
           (pre lg){a := 10}{b := 20}
         else
           (pre lg){c := 30}{d := 40};
           
  cex2 = (lg.a = 0 or lg.c = 0);
  --%PROPERTY cex2;
tel;
