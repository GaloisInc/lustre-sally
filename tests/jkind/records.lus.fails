TODO: See #4 on Github
