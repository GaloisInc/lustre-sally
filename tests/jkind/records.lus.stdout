Sally Model
===========

(define-state-type
   S
   (
     (cex1 Bool)
     (cex2 Bool)
     (delta1 Int)
     (delta2 Int)
     (lemma Bool)
     (ok1 Bool)
     (|gal-initializing| Bool)
     (|lg.a:init| Bool)
     (|lg.a| Int)
     (|lg.b:init| Bool)
     (|lg.b| Int)
     (|lg.c:init| Bool)
     (|lg.c| Int)
     (|lg.d:init| Bool)
     (|lg.d| Int)
     (|lg.e:init| Bool)
     (|lg.e| Int)
     (|pre_lg.a| Int)
     (|pre_lg.b| Int)
     (|pre_lg.c| Int)
     (|pre_lg.d| Int)
     (|pre_lg.e:1| Int)
     (|pre_lg.e| Int)
     (|pre_wp1.p.x| Int)
     (|pre_wp1.p.y| Int)
     (|pre_wp1.weight| Real)
     (|pre_wp2.p.x| Int)
     (|pre_wp2.p.y| Int)
     (|pre_wp2.weight| Real)
     (|wp1.p.x:init| Bool)
     (|wp1.p.x| Int)
     (|wp1.p.y:init| Bool)
     (|wp1.p.y| Int)
     (|wp1.weight:init| Bool)
     (|wp1.weight| Real)
     (|wp2.p.x:init| Bool)
     (|wp2.p.x| Int)
     (|wp2.p.y:init| Bool)
     (|wp2.p.y| Int)
     (|wp2.weight:init| Bool)
     (|wp2.weight| Real))
   (
     (delta1 Int)
     (delta2 Int)))
(define-transition-system
   TS
   S
   (and
      (= |gal-initializing| true)
      (= |wp1.weight:init| true)
      (= |wp1.p.x:init| true)
      (= |wp1.p.y:init| true)
      (= |wp2.weight:init| true)
      (= |wp2.p.x:init| true)
      (= |wp2.p.y:init| true)
      (= |lg.a:init| true)
      (= |lg.c:init| true)
      (= |lg.b:init| true)
      (= |lg.d:init| true)
      (= |lg.e:init| true))
   (and
      (= next.|gal-initializing| false)
      (= next.delta1 input.delta1)
      (= next.delta2 input.delta2)
      (= next.|pre_wp2.weight| state.|wp2.weight|)
      (ite
         state.|wp1.weight:init|
         (= next.|wp1.weight| 1.0)
         (= next.|wp1.weight| next.|pre_wp2.weight|))
      (= next.|wp1.weight:init| false)
      (= next.|pre_wp1.p.x| state.|wp1.p.x|)
      (ite
         state.|wp1.p.x:init|
         (= next.|wp1.p.x| 0)
         (=
            next.|wp1.p.x|
            (+ next.|pre_wp1.p.x| 1)))
      (= next.|wp1.p.x:init| false)
      (= next.|pre_wp1.p.y| state.|wp1.p.y|)
      (ite
         state.|wp1.p.y:init|
         (= next.|wp1.p.y| 0)
         (=
            next.|wp1.p.y|
            (+ next.|pre_wp1.p.y| next.delta1)))
      (= next.|wp1.p.y:init| false)
      (= next.|pre_wp1.weight| state.|wp1.weight|)
      (ite
         state.|wp2.weight:init|
         (= next.|wp2.weight| 2.0)
         (= next.|wp2.weight| next.|pre_wp1.weight|))
      (= next.|wp2.weight:init| false)
      (= next.|pre_wp2.p.x| state.|wp2.p.x|)
      (ite
         state.|wp2.p.x:init|
         (= next.|wp2.p.x| 0)
         (=
            next.|wp2.p.x|
            (+ next.|pre_wp2.p.x| 1)))
      (= next.|wp2.p.x:init| false)
      (= next.|pre_wp2.p.y| state.|wp2.p.y|)
      (ite
         state.|wp2.p.y:init|
         (= next.|wp2.p.y| 20)
         (=
            next.|wp2.p.y|
            (+ next.|pre_wp2.p.y| next.delta2)))
      (= next.|wp2.p.y:init| false)
      (=
         next.ok1
         (or
            (not
               (= next.|wp1.weight| next.|wp2.weight|))
            (not
               (= next.|wp1.p.x| next.|wp2.p.x|))
            (not
               (= next.|wp1.p.y| next.|wp2.p.y|))))
      (=
         next.cex1
         (not
            (and
               (= next.|wp1.p.x| next.|wp2.p.x|)
               (= next.|wp1.p.y| next.|wp2.p.y|))))
      (= next.|pre_lg.a| state.|lg.a|)
      (ite
         state.|lg.a:init|
         (= next.|lg.a| 0)
         (=
            next.|lg.a|
            (ite
               (= next.delta1 0)
               10
               next.|pre_lg.a|)))
      (= next.|lg.a:init| false)
      (= next.|pre_lg.c| state.|lg.c|)
      (ite
         state.|lg.c:init|
         (= next.|lg.c| 0)
         (=
            next.|lg.c|
            (ite
               (= next.delta1 0)
               next.|pre_lg.c|
               30)))
      (= next.|lg.c:init| false)
      (=
         next.cex2
         (or
            (= next.|lg.a| 0)
            (= next.|lg.c| 0)))
      (=
         next.lemma
         (not
            (= next.|wp1.weight| next.|wp2.weight|)))
      (= next.|pre_lg.b| state.|lg.b|)
      (ite
         state.|lg.b:init|
         (= next.|lg.b| 0)
         (=
            next.|lg.b|
            (ite
               (= next.delta1 0)
               20
               next.|pre_lg.b|)))
      (= next.|lg.b:init| false)
      (= next.|pre_lg.d| state.|lg.d|)
      (ite
         state.|lg.d:init|
         (= next.|lg.d| 0)
         (=
            next.|lg.d|
            (ite
               (= next.delta1 0)
               next.|pre_lg.d|
               40)))
      (= next.|lg.d:init| false)
      (= next.|pre_lg.e:1| state.|lg.e|)
      (= next.|pre_lg.e| state.|lg.e|)
      (ite
         state.|lg.e:init|
         (= next.|lg.e| 0)
         (=
            next.|lg.e|
            (ite
               (= next.delta1 0)
               next.|pre_lg.e:1|
               next.|pre_lg.e|)))
      (= next.|lg.e:init| false)))

(query
   TS
   (or |gal-initializing| cex2))
(query
   TS
   (or |gal-initializing| ok1))
(query
   TS
   (or |gal-initializing| lemma))
(query
   TS
   (or |gal-initializing| cex1))

[Lustre] Validating properties:
[Lustre] Property cex2...
  [Sally][Invalid]
[Trace] 
cex2:
 Step | delta1 | delta2 | |-> 
------+--------+--------+-----
 1    | 0      | 0      |     
 2    | 0      | -20    |     
 3    | -1     | 0      |     

[Lustre] Property ok1...
  [Sally][Unknown]
  [Sally][Unknown]
[Lustre] Property lemma...
  [Sally][Valid] using 1 past state
[Lustre] Property cex1...
  [Sally][Invalid]
[Trace] 
cex1:
 Step | delta1 | delta2 | |-> 
------+--------+--------+-----
 1    | 0      | 0      |     
 2    | 0      | -20    |     

[Lustre] Summary:
  [Valid] 1
  [Unknown] 1
  [Invalid] 2
[Lustre] Model status: [Invalid]
